### DISCUSSION TITLE
Time Series Cross Validation

### DISCUSSION THREAD

**Discussion by ngupta23** [pycaret/pycaret#1761] [FAQs]
# Background

Simple random splits when performing time series cross validation is not a good idea for many reasons

1. It does not maintain the temporal dependence in the data that can be important for modeling
2. It can cause leakage by using future (unknown at the time of training) data during the training process.

To solve this problem, well known temporal cross validation techniques such as expanding/rolling window cross-validation and sliding window cross validation have been proposed. [This link](https://topepo.github.io/caret/data-splitting.html) has a good visualization of these techniques (see image below).

![image](https://user-images.githubusercontent.com/33585645/138573429-4255b0ae-332a-4c81-96a8-93cbd18c9905.png)

# Solution in `pycaret`

The `pycaret` time series module allows you to create experiments using both these techniques and provides handy visualization techniques using `plotly`. Lets see how.

NOTE: Jupyter Notebook for this FAQ can be **[found here](https://nbviewer.org/github/ngupta23/pycaret_faqs/blob/main/time_series/pycaret_ts_ttsplit_cv.ipynb)**.

## Common Setup

```python
#### Install library (one time) ----
!pip install pycaret-ts-alpha -U

#### Load data ----
from pycaret.datasets import get_data
y = get_data("airline")
```

## Expanding Window Cross Validation

```python
#### Setup experiment ----
from pycaret.internal.pycaret_experiment import TimeSeriesExperiment
exp = TimeSeriesExperiment()

# Default Setup has 3 folds
# Also, expanding Window CV is the default so it does not need to be explicitly specified.
exp.setup(data=y, fh=12, session_id=42)  

#### Check CV Setup ----
exp.plot_model(plot="cv")
```

![image](https://user-images.githubusercontent.com/33585645/138573523-ac9101d5-15cf-495d-b203-97a1a16eb29b.png)

## Sliding Window Cross Validation

```python
#### Setup experiment ----
from pycaret.internal.pycaret_experiment.time_series_experiment import TimeSeriesExperiment
exp = TimeSeriesExperiment()

# Default Setup has 3 folds
exp.setup(data=y, fh=12, fold_strategy="sliding", session_id=42)

#### Check CV Setup ----
exp.plot_model(plot="cv")
```

![image](https://user-images.githubusercontent.com/33585645/138573555-c735a944-3550-4567-84e2-fdff37902767.png)

## Customization

Users can also fully customize the cross validation using custom `sktime` splitters. For example, a user might want to use a sliding window splitter but with different settings than that used in `pycaret` by default. 

```python
#### Setup experiment ----
exp = TimeSeriesExperiment()

#### Add a custom Cross Validation Splitter with a
####  Step length = 6, Forecast Horizon = 24, Window Length = 48
my_custom_cv = SlidingWindowSplitter(fh=np.arange(1, 25), window_length=48, step_length=6)
exp.setup(data=y, session_id=42, fold_strategy=my_custom_cv)

#### Check CV Setup on the Training Split ----
exp.plot_model(plot="cv")
```

![image](https://user-images.githubusercontent.com/33585645/142192171-ff69704d-0e02-4ede-8bd9-025ae57c8573.png)

**Comment by moezali1** [comment]
PyCaret 4.0 sunset the legacy time-series module (sktime/prophet integration), so this FAQ no longer maps to a supported code path.

If you still need time-series forecasting, use sktime / prophet / statsforecast directly — those projects have surpassed what the PyCaret wrapper offered.

For 4.0 docs, see https://pycaret.org/docs and the cleanup notice at https://github.com/pycaret/pycaret/issues/4177.

Closing as outdated.
